Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs

```diff
@@ -22,230 +26,370 @@ namespace SS3D.Systems.Tile.TileMapCreator
                 {
-                    return;
+                    _tileSystem.RpcClearTileObject(_selectedObject.nameString, buildGhost.TargetPosition, buildGhost.Dir);
                 }
-
-                _tileSystem.RpcPlaceObject(_selectedObject.nameString, snappedPosition, _ghostManager.Dir, replaceExisting);
-                RefreshGhost();
             }
         }
-
-        [ServerOrClient]
-        private void FindAndDeleteItem(Vector3 worldPosition)
+        
+        private void FindAndDeleteItem()
         {
             Ray ray = Camera.main.ScreenPointToRay(Input.mousePosition);
-
             if (Physics.Raycast(ray, out RaycastHit hitInfo))
             {
                 PlacedItemObject placedItem = hitInfo.collider.gameObject.GetComponent<PlacedItemObject>();
-
                 if (placedItem != null)
                 {
-                    _tileSystem.RpcClearItemObject(placedItem.NameString, worldPosition);
+                    _tileSystem.RpcClearItemObject(placedItem.NameString, placedItem.gameObject.transform.position);
                 }
             }
         }
+        #endregion
 
-        [ServerOrClient]
-        private Vector3 GetMousePosition()
+        #region Drag algorithms
+        private void LineDrag(Vector3 position)
         {
-            Ray ray = Camera.main.ScreenPointToRay(Input.mousePosition);
-
-            if (_plane.Raycast(ray, out float distance))
+            foreach (Vector3 tile in FindTilesOnLine(_dragStartPostion, position))
             {
-                return ray.GetPoint(distance);
+                BuildGhost buildGhost = CreateGhost(_selectedObject.prefab, tile, _direction);
+                buildGhost.TargetPosition = tile;
+                RefreshGhost(buildGhost);
+                _buildGhosts.Add(buildGhost);
             }
+        }
 
-            return Vector3.zero;
+        private void SquareDrag(Vector3 position)
+        {
+            int x1 = (int)Math.Min(_dragStartPostion.x, position.x);
+            int x2 = (int)Math.Max(_dragStartPostion.x, position.x);
+            int y1 = (int)Math.Min(_dragStartPostion.z, position.z);
+            int y2 = (int)Math.Max(_dragStartPostion.z, position.z);
+
+            for (int i = y1; i <= y2; i++)
+            {
+                for (int j = x1; j <= x2; j++)
+                {
+                    Vector3 tile = new(j, 0, i);
+                    BuildGhost buildGhost = CreateGhost(_selectedObject.prefab, tile, _direction);
+                    buildGhost.TargetPosition = tile;
+                    RefreshGhost(buildGhost);
+                    _buildGhosts.Add(buildGhost);
+                }
+            }
+        }
+        /// <summary>
+        /// Bernstein algorithm for drawing lines in a grid.
+        /// </summary>
+        /// <returns>Tiles, that line passes through</returns>
+        private Vector3[] FindTilesOnLine(Vector3 firstPoint, Vector3 secondPoint)
```

Comment:
this is cool, and could be useful elsewhere. I think it belongs in a static class like "MathHelper" or another one in the utility namespace.